cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/DefaultHttpServiceResponder.java

```diff
@@ -183,4 +191,36 @@ private void emitMetrics(int status) {
     metricsCollector.increment(builder.toString(), 1);
     metricsCollector.increment("requests.count", 1);
   }
+
+  private static final class BufferedResponse {
+
+    private final int status;
+    private final ChannelBuffer channelBuffer;
+    private final String contentType;
+    private final Multimap<String, String> headers;
+
+    private BufferedResponse(int status, ByteBuffer content,
+                             String contentType, Multimap<String, String> headers) {
+      this.status = status;
+      this.channelBuffer = content == null ? null : ChannelBuffers.wrappedBuffer(ByteBuffers.copy(content));
```

Comment:
Actually for sendJson and sendString cases, you don't need to copy the `ByteBuffer`. Maybe it's better not to do copy in here but leave it to the one who call this constructor to decide whether copying is needed or not.